tests/ostree_mock.c

```diff
@@ -3,26 +3,137 @@
   }
+
   return orig(self);
 }
 
-const char* ostree_deployment_get_bootcsum(OstreeDeployment* self) {
-  char* (*orig)(OstreeDeployment*) = dlsym(RTLD_NEXT, __func__);
-  char mocked_name[100];
-  snprintf(mocked_name, sizeof(mocked_name), "%s_mock", __func__);
-  char* (*mocked)(OstreeDeployment*) = dlsym(RTLD_DEFAULT, mocked_name);
+struct OstreeDeployment {
+  int serial;
+  char rev[255];
+};
+
+static int read_deployment_info(FILE* file, struct OstreeDeployment* deployment) {
+  struct OstreeDeployment deployment_res;
+
+  // TODO: make it more robust and reliable as a file might contain a value exceeding 255
+  if (fscanf(file, "%s\n", deployment_res.rev) != 1) {
+    return -1;
+  }
+
+  // TODO: make it more robust and reliable as a file might contain a value exceeding 255 and/or char(s) instead of a
+  // number for the serial value
+  if (fscanf(file, "%d\n", &deployment_res.serial) != 1) {
+    return -1;
+  }
+
+  *deployment = deployment_res;
+  return 0;
+}
+
+static int get_ostree_deployment_info(const char* filename, struct OstreeDeployment* deployed,
+                                      struct OstreeDeployment* booted) {
+  printf("\n>>>>>>>>>> Reading the file: %s\n\n", filename);
+  FILE* file = fopen(filename, "r");
+  if (file == NULL) {
+    printf("\n>>>>>>>>>> Failed to open the file: %s\n\n", filename);
+    return -1;
+  }
+
+  int return_code = -1;
+
+  while (1) {
+    struct OstreeDeployment deployed_res;
+    if (read_deployment_info(file, &deployed_res) != 0) {
+      break;
+    }
+
+    if (booted != NULL) {
+      struct OstreeDeployment booted_res;
+      if (read_deployment_info(file, &booted_res) != 0) {
+        break;
+      }
+      *booted = booted_res;
+    }
+
+    *deployed = deployed_res;
+    return_code = 0;
+    break;
+  }
+
+  fclose(file);
+  if (return_code != 0) {
+    printf("%s: Failed to read the deployment info from: %s\n", __FILE__, filename);
+  }
+  return return_code;
+}
+
+static struct OstreeDeployment g_deployed;
+
+const char* ostree_deployment_get_csum(OstreeDeployment* self) {
+  const char* (*orig)(OstreeDeployment*) = (const char* (*)(OstreeDeployment*))(dlsym(RTLD_NEXT, __func__));
 
-  if (mocked) {
-    return mocked(self);
+  const char* deployment_version_file = NULL;
+
+  if ((deployment_version_file = getenv(OSTREE_DEPLOYMENT_VERSION_FILE)) != NULL) {
+    if (get_ostree_deployment_info(deployment_version_file, &g_deployed, NULL) != 0) {
+      return NULL;
+    }
+
+    return g_deployed.rev;
   }
+
   return orig(self);
 }
+
+OstreeDeployment* ostree_sysroot_get_booted_deployment_from_file(const char* filename) {
+  struct OstreeDeployment deployed;
+  struct OstreeDeployment booted;
+
+  if (get_ostree_deployment_info(filename, &deployed, &booted) != 0) {
+    return NULL;
+  }
+
+  OstreeDeployment* new_depl =
+      ostree_deployment_new(0, OSTREE_DEPLOYMENT_OS_NAME, deployed.rev, deployed.serial, booted.rev, booted.serial);
+  return new_depl;
+}
```

Comment:
So who is using this version? And what exactly is it doing that the old version couldn't do? I'm having trouble following it.

Reply:
From the PR description, it could be useful for plugging the mocks in other tests apparently coming in other PRs. The downside is that there is nothing guaranteeing that it works in the current state.

Reply:
> So who is using this version?
This version of the ostree mock is being used by one of the `test_backend_failure` test suite, specifically the test that test an ostree update in case of download failures. The test is currently under development, it will go as a second commit into this PR.

> And what exactly is it doing that the old version couldn't do?
The old version implies that the `mock` function is actually implemented and exposed/exported which is require from an user to implement the `mock` in some C/C++ file. This fact does not allow (or make it not feasible) to reuse it from more complex integration tests written in other languages, e.g test_backend_failure, or in case of standalone running of aktualizr on a local host configured with the ostree package manager.

The given version does not require writing any mock function from an user, just set LD_PRELOAD environment variable for aktualizr process and control/manage the ostree deployed and booted revisions via a file from other processes/scripts which makes it universally applicable in cases when aktualizr is executed in a separate process.